### PR TITLE
Fix for #422: originate sessions from the interface matching the router ID

### PR DESCRIPTION
**Pull request by briantopping**
See #422

### CODE REVIEW

File: internal/bgp/bgp.go [hunk 1]
```diff
@@ -168,7 +168,15 @@ func (s *Session) connect() error {
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 	deadline, _ := ctx.Deadline()
-	conn, err := dialMD5(ctx, s.addr, s.password)
+
+	laddr := "[::]:0"
+	if (s.routerID.To4() != nil) {
+		laddr = fmt.Sprintf("%s:0", s.routerID.String())
+	} else if (s.routerID.To16() != nil) {
```
Comment (danderson) [inline internal/bgp/bgp.go]: Router IDs cannot be IPv6 addresses in BGP, so I think you can remove this branch. `s.routerID != nil` should imply `To4() != nil`

File: internal/bgp/bgp.go [hunk 2]
```diff
@@ -168,7 +168,15 @@ func (s *Session) connect() error {
...
+		laddr = fmt.Sprintf("[%s]:0", s.routerID.String())
```
Comment (danderson) [inline internal/bgp/bgp.go]: The BGP spec doesn't require the router ID to match any local interfaces. What happens if I specify a router ID of 1.2.3.4, and that address doesn't exist locally? I think there needs to be another check here, to see if the router ID is actually configured on a local interface (`net.Interfaces()` lets you find that out).

**Comment by briantopping** [comment]
Thanks @danderson, the problem arose in peering with BIRD 2.0.2 ( I didn’t test other versions) where BIRD would get extremely confused about upstream routes if the source address of the socket conflicted with another router that was using that address as a router ID. 

I filed the issue with BIRD as well (seems this is done on their users list). The response there was a mild irritation about addresses that are not on interfaces. I think there is some merit to this perspective. I spent some time with the BIRD source, but it’s Autotools-based and CLion was having none of it. It uses a bison/flex stack to execute their filter language and I simply wasn’t able to crack the parse tree executor to figure out where a fix was on that end. 

You may be correct that these additional lines are unnecessary to interoperate with vendor gear that does not run containers on the same node, but it’s a slightly more challenging situation when routes are being sourced from the same hardware that the routing protocol instance is running on. Cisco has a commodity hardware router stack, but I don’t have a license to test it. 

It may be that this patch is less concerning behind a feature flag. I’m happy to add that if you wish.

**Comment by briantopping** [comment]
Was there something about the PR that was out-of-spec or was the fix applied without this source? Just curious why it was closed with no comment.

**Comment by danderson** [comment]
Sorry, I was going through inactive PRs and closing them. I assumed you'd dropped this, clearly mistakenly!

Reading back, I didn't think you were blocked on anything, but it sounds like I'm reading wrong. Sorry! What do you need from me to get this submitted?

**Comment by briantopping** [comment]
No worries! 

I don’t remember what I solved with it, I was just trying to figure out if there was some problem with the PR or it was already applied. Was mostly trying to help others avoid the same trap.

**Comment by briantopping** [comment]
Not exactly sure, I will have to look through the branch.

**Pull request closed without merging**